src/database/postgres.go

```diff
@@ -0,0 +1,198 @@
+package database
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"time"
+
+	pb "accretional.com/semantifly/proto/accretional.com/semantifly/proto"
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgconn"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type PgxIface interface {
+	Begin(context.Context) (pgx.Tx, error)
+	Close(context.Context) error
+	Query(ctx context.Context, sql string, args ...interface{}) (pgx.Rows, error)
+	Exec(ctx context.Context, sql string, arguments ...interface{}) (pgconn.CommandTag, error)
+}
+
+func establishConnection(ctx context.Context) (*pgx.Conn, error) {
+	conn, err := pgx.Connect(ctx, os.Getenv("DATABASE_URL"))
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect to database: %w", err)
+	}
+	return conn, nil
+}
+
+func initializeDatabaseSchema(ctx context.Context, conn PgxIface) error {
+    tx, err := conn.Begin(ctx)
+    if err != nil {
+        return fmt.Errorf("unable to connect to database: %w", err)
+    }
+    defer tx.Rollback(ctx)
+
+    // Create the main table
+    _, err = tx.Exec(ctx, `
+        CREATE TABLE IF NOT EXISTS index_list (
```

Comment:
I think ideally we would be able to use reflection or something to automatically generate the column names/types from some input type, but that could be quite complicated. I do think we can and should do better than hardcoding the table columns though - it works against a lot of the benefits of using protobufs in the first place if we have to start worrying about forward/backwards compatability and need to maintain two different versions of the same type

 Can you use something like https://github.com/mpartel/postgres-protobuf to insert the protobuf to the database with the following schema?

name TEXT PRIMARY KEY,
entry JSONB, // automatically generated from https://github.com/mpartel/postgres-protobuf
// only the specific fields that do not work properly if nested within entry. Probably vector embeddings in the future, but not yet

If you have a better approach I'm open to talking about it but I really don't think we should hardcode the schema like this

Reply:
I was more concerned about the performance of the query if we store the entire entry as a JSONB. 

> Queries need to load and scan through the entire protobuf column. Storing and querying large protobufs can be significantly slower than splitting the data into columns that can be queried individually.

I will look more into this.

Reply:
Not using postgres-protobuf as it's not very flexible. Instead, converting protobuf to JSONB using protojson and then storing in the tables.